server/src/api/api.ts

```diff
@@ -43,465 +24,572 @@ async function getItem(itemId: number, isAdmin: boolean): Promise<Item|null> {
-     * Access level: any signed in user
-     * @param root
-     * @param _args
-     * @param _context
-     */
-    item: async (root, _args, _context): Promise<Item|null> => {
-        // @ts-ignore
-        const {args, context} = fixArguments(root, _args, _context);
-
-        return await getItem(args.id, context.user.admin);
-    },
-    /**
-     * Bulk items API
-     * TODO: pagination/returned quantity limit
-     * Access level: any signed in user
-     * @param root
-     * @param _args
-     * @param _context
-     */
-    items: async (root, _args, _context): Promise<Item[]> => {
-        // @ts-ignore
-        const {args, context} = fixArguments(root, _args, _context);
-        const items = await DB.from("items")
-            .join("categories", "items.category_id", "=", "categories.category_id");
-
-        return items.map(item => {
-            return {
-                ...item,
-                id: item.item_id,
-                category: item.category_name,
-                price: onlyIfAdmin(item.price, context.user.admin),
-                owner: onlyIfAdmin(item.owner, context.user.admin)
-            };
-        });
-    },
-    categories: async (root, _args, _context): Promise<Category[]> => {
-        // @ts-ignore
-        const {args, context} = fixArguments(root, _args, _context);
-        return await DB.from("categories");
-    },
-    requests: async (root, _args, _context): Promise<Request[]> => {
-        // @ts-ignore
-        const {args, context} = fixArguments(root, _args, _context);
+            console.log(qtyNotAtDeskObj);
 
-        const searchObj: any = {};
+            const qtyPartialReserved = await DB.from("requests")
+                .whereIn("status", ["SUBMITTED", "APPROVED", "READY_FOR_PICKUP"])
+                .groupBy("request_item_id")
+                .orderBy("request_item_id")
+                .select("request_item_id")
+                .count("request_id");
 
+            const qtyPartialReservedObj = {};
 
-        if (args.search.item_id) {
-            searchObj.item_id = args.search.item_id;
-        }
+            qtyPartialReserved.forEach((value, index, arr) => {
+                qtyPartialReservedObj[value.request_item_id.toString()] = parseInt(value.count, 10);
+            });
 
-        if (args.search.request_id) {
-            searchObj.request_id = args.search.request_id;
-        }
+            console.log("qtyPartialReserved", qtyPartialReservedObj);
+
+
+            return items.map(item => {
+                const totalAvailable: number = item.totalAvailable;
+                console.log(totalAvailable);
+                const qtyUnreserved: number = totalAvailable - ((qtyNotAtDeskObj[item.item_id.toString()] || 0) + (qtyPartialReservedObj[item.item_id.toString()] || 0));
+                const qtyInStock: number = totalAvailable - (qtyNotAtDeskObj[item.item_id.toString()] || 0);
+                console.log(qtyUnreserved, qtyInStock);
```

Comment:
did we agree on showing users just "can reserve" or "OOS"?

Reply:
yes